File: 02_TodoList/TodoList.java
import java.util.ArrayList;
import java.util.Scanner;

/**
 * TodoList - Simple task management system
 * Add, view, complete, and delete tasks
 */
public class TodoList {
    private static ArrayList<String> tasks = new ArrayList<>();
    private static ArrayList<Boolean> completed = new ArrayList<>();

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        boolean running = true;

        while(running) {
            System.out.println("\n=== Todo List ===");
            System.out.println("1. Add task");
            System.out.println("2. View tasks");
            System.out.println("3. Complete task");
            System.out.println("4. Delete task");
            System.out.println("5. Exit");
            System.out.print("Choose option: ");

            int choice = scanner.nextInt();
            scanner.nextLine(); // consume newline

            switch(choice) {
                case 1:
                    System.out.print("Enter task: ");
                    String task = scanner.nextLine();
                    tasks.add(task);
                    completed.add(false);
                    System.out.println("Task added!");
                    break;
                case 2:
                    viewTasks();
                    break;
                case 3:
                    System.out.print("Enter task number to complete: ");
                    int completeIdx = scanner.nextInt() - 1;
                    if(completeIdx >= 0 && completeIdx < tasks.size()) {
                        completed.set(completeIdx, true);
                        System.out.println("Task completed!");
                    } else {
                        System.out.println("Invalid task number!");
                    }
                    break;
                case 4:
                    System.out.print("Enter task number to delete: ");
                    int deleteIdx = scanner.nextInt() - 1;
                    if(deleteIdx >= 0 && deleteIdx < tasks.size()) {
                        tasks.remove(deleteIdx);
                        completed.remove(deleteIdx);
                        System.out.println("Task deleted!");
                    } else {
                        System.out.println("Invalid task number!");
                    }
                    break;
                case 5:
                    running = false;
                    System.out.println("Goodbye!");
                    break;
                default:
                    System.out.println("Invalid option!");
            }
        }
        scanner.close();
    }

    private static void viewTasks() {
        if(tasks.isEmpty()) {
            System.out.println("No tasks!");
            return;
        }
        System.out.println("\nYour tasks:");
        for(int i = 0; i < tasks.size(); i++) {
            String status = completed.get(i) ? "[✓]" : "[ ]";
            System.out.println((i + 1) + ". " + status + " " + tasks.get(i));
        }
    }
}

File: Dart/003_TodoList/todolist.dart
import 'dart:io';

class TodoItem {
  String task;
  bool completed;

  TodoItem(this.task, {this.completed = false});

  @override
  String toString() => completed ? '[✓] $task' : '[ ] $task';
}

class TodoList {
  List<TodoItem> items = [];

  void addTask(String task) {
    items.add(TodoItem(task));
    print('Added: $task');
  }

  void completeTask(int index) {
    if (index >= 0 && index < items.length) {
      items[index].completed = true;
      print('Completed: ${items[index].task}');
    }
  }

  void displayTasks() {
    if (items.isEmpty) {
      print('No tasks yet!');
      return;
    }

    print('\n=== Todo List ===');
    for (int i = 0; i < items.length; i++) {
      print('$i. ${items[i]}');
    }
  }
}

void main() {
  TodoList todoList = TodoList();

  while (true) {
    print('\n1. Add Task');
    print('2. Complete Task');
    print('3. View Tasks');
    print('4. Exit');
    stdout.write('Choose option: ');

    String? choice = stdin.readLineSync();

    switch (choice) {
      case '1':
        stdout.write('Enter task: ');
        String? task = stdin.readLineSync();
        if (task != null && task.isNotEmpty) {
          todoList.addTask(task);
        }
        break;
      case '2':
        stdout.write('Task number to complete: ');
        int index = int.parse(stdin.readLineSync()!);
        todoList.completeTask(index);
        break;
      case '3':
        todoList.displayTasks();
        break;
      case '4':
        print('Goodbye!');
        return;
    }
  }
}
